### PR TITLE
Port to ring

### PR DESCRIPTION
**Pull request by romac**

### CODE REVIEW

File: src/hashutils.rs [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    /// Compute the hash of the concatenation of `left` and `right`
+    fn combine_hashes(&'static self, left: &[u8], right: &[u8]) -> Vec<u8>;
+}
+
+impl HashUtils for Algorithm {
+
+    fn hash_bytes(&'static self, bytes: &[u8]) -> Vec<u8> {
+        let mut context = Context::new(self);
+        context.update(bytes);
+        context.finish().as_ref().into()
```
Comment (briansmith) [inline src/hashutils.rs:20]: FYI, you can just use digest::digest(self, bytes).as_ref().into()

Comment (briansmith) [reply src/hashutils.rs:20]: In fact, you don't need to convert to `Vec<u8>`. You can return the `Digest` directly from this and the other function, saving a malloc each.

File: src/hashutils.rs [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    }
+
+    fn combine_hashes(&'static self, left: &[u8], right: &[u8]) -> Vec<u8> {
+        let mut context = Context::new(self);
+        let combined = [left, right].concat();
```
Comment (briansmith) [inline src/hashutils.rs:25]: Surely it is faster to do this:
```rust
context.update(left);
context.update(right);
```

**Comment by FredericJacobs** [comment]
Thanks @briansmith for the feedback! We're working on addressing those points.

**Pull request merged**